dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,12 +607,658 @@ class FunctionsJsonArray : public IFunction
+{
+public:
+    static constexpr auto name = "cast_decimal_as_json";
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionCastDecimalAsJson>(); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForNulls() const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if unlikely (!arguments[0]->isDecimal())
+            throw Exception(
+                fmt::format("Illegal type {} of argument of function {}", arguments[0]->getName(), getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        auto col_to = ColumnString::create();
+        auto & data_to = col_to->getChars();
+        JsonBinary::JsonBinaryWriteBuffer write_buffer(data_to);
+        auto & offsets_to = col_to->getOffsets();
+        auto rows = block.rows();
+        offsets_to.resize(rows);
+
+        const auto & from = block.getByPosition(arguments[0]);
+        TypeIndex from_type_index = from.type->getTypeId();
+        switch (from_type_index)
+        {
+        case TypeIndex::Decimal32:
+            doExecute<Decimal32>(write_buffer, offsets_to, from.column);
+            break;
+        case TypeIndex::Decimal64:
+            doExecute<Decimal64>(write_buffer, offsets_to, from.column);
+            break;
+        case TypeIndex::Decimal128:
+            doExecute<Decimal128>(write_buffer, offsets_to, from.column);
+            break;
+        case TypeIndex::Decimal256:
+            doExecute<Decimal256>(write_buffer, offsets_to, from.column);
+            break;
+        default:
+            throw Exception(
+                fmt::format(
+                    "Illegal type {} of argument of function {}",
+                    magic_enum::enum_name(from_type_index),
+                    getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+        data_to.resize(write_buffer.count());
+        block.getByPosition(result).column = std::move(col_to);
+    }
+
+private:
+    template <typename FromType>
+    static void doExecute(
+        JsonBinary::JsonBinaryWriteBuffer & data_to,
+        ColumnString::Offsets & offsets_to,
+        const ColumnPtr & column_ptr_from)
+    {
+        const auto * column_from = checkAndGetColumn<ColumnDecimal<FromType>>(column_ptr_from.get());
+        RUNTIME_CHECK(column_from);
+        for (size_t i = 0; i < column_from->size(); ++i)
+        {
+            const auto & field = (*column_from)[i].template safeGet<DecimalField<FromType>>();
+            JsonBinary::appendNumber(data_to, static_cast<Float64>(field));
```

Comment:
Decimal is append as a float?

Reply:
No decimal type in json.

Reply:
yes, same as https://github.com/pingcap/tidb/blob/90628349860718bb84c94fe7dc1e1f9bd9da4348/pkg/expression/builtin_cast.go#L854-L865